python/tank/platform/engine.py

```diff
@@ -452,6 +487,95 @@ def destroy_engine(self):
         Implemented by deriving classes.
         """
         pass
+
+    def change_context(self, new_context):
+        """
+        Called when the engine is being asked to change contexts. This
+        will only be allowed if the engine explicitly suppose on-the-fly
+        context changes by way of its context_change_allowed property. Any
+        apps that do not support context changing will be restarted instead.
+        Custom behavior at the engine level should be handled by overriding
+        one or both of pre_context_change and post_context_change methods.
+
+        :param new_context:     The context to change to.
+        """
+        # Make sure we're allowed to change context at the engine level.
+        if not self.context_change_allowed:
+            self.log_debug("Engine %s does not allow context changes." % self.name)
+            raise TankContextChangeNotAllowedError()
+
+        # Make sure that this engine is configured to run in the new context,
+        # and that it's the EXACT same engine. This can be handled by comparing
+        # the current engine's descriptor to the one coming from the new environment.
+        # If this fails then it's more than just the engine not supporting the
+        # context change, it's that the target context isn't configured properly.
+        # As such, we'll let any exceptions (mostly TankEngineInitError) bubble
+        # up since it's a critical error case.
+        (new_env, engine_descriptor) = _get_env_and_descriptor_for_engine(
+            engine_name=self.instance_name,
+            tk=self.tank,
+            context=new_context,
+        )
+
+        # Make sure that the engine in the target context is the same as the current
+        # engine. If the descriptors are the same object then we know for sure we're
+        # good to go. Otherwise we need to do a little more work before we can feel
+        # safe to continue.
+        if engine_descriptor is not self.descriptor:
+            self.log_debug("Engine %s does not match descriptors between %s and %s." % (
+                self.instance_name,
+                str(self.context),
+                str(new_context)
+            ))
+            if engine_descriptor.get_path() == self.descriptor.get_path():
+                # If the descriptors are different but the install location for
+                # the engine remains the same, then we can feel pretty safe. This
+                # is likely due to the engine running off of a dev descriptor.
+                self.log_debug("Engine descriptor disk paths match, allowing context change.")
```

Comment:
actually, now that i am thinking about this a bit more, maybe it's better that this logic is all in the eq operator for the descriptor. Then we don't need to take dev descriptors etc into account in the engine - that's all encapsulated within the descriptor system. 

I think maybe for dev descriptors, the equality op should always return False between two instances? And the engine should always restart, assuming worst case scenario (e.g. that someone did a change to the code somewhere)? What do you think?